components/txn_types/src/lock.rs

```diff
@@ -461,126 +641,180 @@ impl Lock {
-                lock.into_owned().into_lock_info(raw_key),
-            )));
+impl SharedLocks {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn new_with_txn_infos(
+        txn_info_segments: HashMap<TimeStamp, Either<Vec<u8>, Lock>>,
+    ) -> Self {
+        Self { txn_info_segments }
+    }
+
+    pub fn parse(mut b: &[u8]) -> Result<Self> {
+        if b.is_empty() {
+            return Err(Error::from(ErrorInner::BadFormatLock));
+        }
+        let lock_type = LockType::from_u8(b.read_u8()?).ok_or(ErrorInner::BadFormatLock)?;
+        assert!(lock_type == LockType::Shared);
+        if b.is_empty() {
+            return Ok(Self::new());
         }
 
-        if ts == TimeStamp::max()
-            && raw_key == lock.primary
-            && !lock.use_async_commit
-            && !lock.use_one_pc
-        {
-            // When `ts == TimeStamp::max()` (which means to get latest committed version
-            // for primary key), and current key is the primary key, we ignore
-            // this lock.
-            return Ok(());
+        let mut segments = HashMap::default();
+        while !b.is_empty() {
+            match b.read_u8()? {
+                SHARED_LOCK_TXNS_INFO_PREFIX => {
+                    let len = number::decode_var_u64(&mut b)? as usize;
+                    segments.reserve(len + 1); // some schedulers may append a new lock, reserve for it.
+                    for _ in 0..len {
+                        let lock_bytes = bytes::decode_compact_bytes(&mut b)?;
+                        let lock_ts = detect_lock_ts(&lock_bytes)?;
+                        segments.insert(lock_ts, Either::Left(lock_bytes));
+                    }
+                }
+                _ => {
+                    // To support forward compatibility, all fields should be serialized in order
+                    // and stop parsing if meets an unknown byte.
+                    break;
+                }
+            }
         }
 
-        // There is a pending lock. Client should wait or clean it.
-        Err(Error::from(ErrorInner::KeyIsLocked(
-            lock.into_owned().into_lock_info(raw_key),
-        )))
+        Ok(Self::new_with_txn_infos(segments))
     }
 
-    // Check if lock could be bypassed for isolation level `RcCheckTs`.
-    fn check_ts_conflict_rc_check_ts(
-        lock: Cow<'_, Self>,
-        key: &Key,
-        ts: TimeStamp,
-        bypass_locks: &TsSet,
-    ) -> Result<()> {
-        if lock.lock_type == LockType::Lock || lock.is_pessimistic_lock() {
-            // Ignore lock when the lock's type is Lock or Pessimistic.
-            return Ok(());
+    pub fn len(&self) -> usize {
+        self.txn_info_segments.len()
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn contains_start_ts(&self, start_ts: TimeStamp) -> bool {
+        self.txn_info_segments.contains_key(&start_ts)
+    }
+
+    pub fn get_lock(&mut self, ts: &TimeStamp) -> Option<&Lock> {
```

Comment:
Why `mutable self` is needed here? If this function call would consume the underlying memory content, better to name it something like `take_lock` or  `mut_take_lock`.